### PR TITLE
Fix lifetimebound for field access

### LINKED ISSUE #81589: [[lifetimebound]] basic members tracking
Hi!

Would be amazing if the following is caught to:

https://godbolt.org/z/5ze9s5roj
```cpp
#include <utility>

struct Struct {
  std::pair<int, int> member;

  const auto* foo() const [[clang::lifetimebound]] { return &member; }
};

int main() {
  // auto& caught = *Struct().foo();
  auto& not_caught = Struct().foo()->second;  // this is a bug
  (void)not_caught;
}
```

### PR DESCRIPTION
**Pull request by usx95**
Fixes: https://github.com/llvm/llvm-project/issues/81589

There is no way to switch this off without  `-Wno-dangling`.

### CODE REVIEW

File: clang/test/SemaCXX/attr-lifetimebound.cpp [hunk 1]
```diff
@@ -47,6 +47,26 @@ namespace usage_ok {
...
+    };
+    Pair p;  
+    FieldCheck(const int a): p(a){}
+    Pair& getPairR() [[clang::lifetimebound]] { return p; }
+    Pair* getPairP() [[clang::lifetimebound]] { return &p; }
+  };
+  void test_field_access() {
+    const int& a = FieldCheck{0}.getPairR().a; // expected-warning {{temporary bound to local reference 'a' will be destroyed at the end of the full-expression}}
+    const int& b = FieldCheck{0}.getPairP()->b; // expected-warning {{temporary bound to local reference 'b' will be destroyed at the end of the full-expression}}
+    const int& c = FieldCheck{0}.getPairP()->c.a; // expected-warning {{temporary bound to local reference 'c' will be destroyed at the end of the full-expression}}
```
Comment (Xazax-hun) [inline clang/test/SemaCXX/attr-lifetimebound.cpp:68]: Could we add some negative tests? (e.g, when there is no loiftimebound annotation on the method).

**Review by Xazax-hun — APPROVED** [review]
LGTM!

File: clang/test/SemaCXX/attr-lifetimebound.cpp (hunk 1 above)
Comment (usx95) [reply clang/test/SemaCXX/attr-lifetimebound.cpp]: Done.

File: clang/test/SemaCXX/attr-lifetimebound.cpp [hunk 2]
```diff
@@ -47,6 +47,29 @@ namespace usage_ok {
...
   }
+
+  struct FieldCheck {
+    struct Set {
+      int a;
+    };
+    struct Pair {
+      const int& a;
+      int b;
+      Set c;
```
Comment (hokein) [inline clang/test/SemaCXX/attr-lifetimebound.cpp:58]: Can you add a test case of pointer type file, e.g. `int* d;` to ensure that we will not fire a diagnostic on `const int* d = FieldCheck{x}.getR().d;`?

**Review by hokein — COMMENTED** [review]
Thanks.

File: clang/test/SemaCXX/attr-lifetimebound.cpp (hunk 2 above)
Comment (usx95) [reply clang/test/SemaCXX/attr-lifetimebound.cpp:58]: Done.

**Review by hokein — APPROVED** [review]

**Comment by AtariDreams** [comment]
Do we want to backport this fix to 19.x?

**Comment by usx95** [comment]
Makes sense to me. Added cherrypick in https://github.com/llvm/llvm-project/issues/81589

**Comment by Xazax-hun** [comment]
Looks like cherry-pick failed and this needs manual back porting.

**Pull request merged**